### ISSUE TITLE
[FEATURE] Dynamic memory allocation

### ISSUE THREAD

**Issue by frankhart2018** [cimplec/sim-c#33]
**Is your feature request related to a problem? Please describe.**
Dynamic memory allocation in simC.

<strong>Note:</strong> Discuss syntax with a maintainer before implementing.

**Comment by viggi-v** [comment]
Hi,

Is it possible for me to take this issue up?

**Comment by frankhart2018** [comment]
Sure @viggi-v. Assigning this to you. Thanks for showing interest to contribute.

**Comment by viggi-v** [comment]
```
var a = new int[2] // for malloc
or 
var a = new typeof(b)[2] // for malloc

free(a) // for free

reshape(a, newsize) // for realloc
// newsize would be just an integer, the simc compiler would convert it to sizeof(datatype)*count
```
Is this good?

**Comment by dhairyaj** [comment]
for malloc, the first option seems good
free(a) is fine
instead of reshape, use realloc

**Comment by Gaurav7888** [comment]
@frankhart2018 Hi I want to work upon this issue.
will use malloc for this purpose..

**Comment by frankhart2018** [comment]
@Gaurav7888 please mention the syntax here, once that is approved I or any other maintainer can assign you to this.

**Comment by Gaurav7888** [comment]
Like (pointer type ) malloc(sizeof() )
This is in general syntax gonna use .

**Comment by frankhart2018** [comment]
DMA includes allocating, reallocating, and deleting.


@Aayush-99 @PranshulDobriyal

**Comment by Gaurav7888** [comment]
Free() could be used.

**Comment by PranshulDobriyal** [comment]
@Gaurav7888  What syntax do you propose for the same in SimC ?

**Comment by Gaurav7888** [comment]
@PranshulDobriyal  cpython  have some functions to do abv things .

**Comment by PranshulDobriyal** [comment]
@Gaurav7888 Yes that will explain the backend of the desired functions, but what front-end (syntax in simC) do you propose. 
For ex-
front-end (SimC code) -> print("hello world!!")
back-end - various functions that generate the corresponding C code -> printf("hello world!!");

**Comment by Math-O5** [comment]
@Gaurav7888 They are asking for something like:

Example of Sim-C allocate memory:

```
var i = new int
var j = new int[10]
```

It's like SimC will interpreter the command, not the code generated.

**Comment by Math-O5** [comment]
@frankhart2018 Sim-C should predict the type of the variable? If so, it's change a lot of expected code. 

A example.

SimC code:
```
var i = new var
var j = new var[10]
```

**Comment by victorjzsun** [comment]
What do we need dynamic memory allocation for other than variable length arrays?

**Comment by frankhart2018** [comment]
Can you start a discussion for this @victorjzsun.

**Comment by victorjzsun** [comment]
Created a discussion [here](https://github.com/cimplec/sim-c/discussions/442)
